File: UNUD/BuildIT/day2/bubbleSort.dart
import 'dart:io';

List<int> bubbleSort(List<int> source) {
  List<int> toSort = source;

  void swap(int indexA, int indexB) {
    int temp = toSort[indexA];
    toSort[indexA] = toSort[indexB];
    toSort[indexB] = temp;
  }

  while (true) {
    int counter = 0;
    
    for (var i = 0; i < toSort.length - 1; i++) {
      if (toSort[i] > toSort[i + 1]) {
        counter++;
        swap(i, i + 1);
      }
    }

    if (counter == 0) {
      return toSort;
    }
  }
}

void main(List<String> args) {
  List<int> src = [1, 5, 3, 2, 5, 7, 2];

  for (var i = 0; i < src.length; i++) {
    stdout.write("${src[i]} ");
  }

  print("");
  var sorted = bubbleSort(src);
  print("sorted");

  for (var i = 0; i < sorted.length; i++) {
    stdout.write("${sorted[i]} ");
  }
}

File: UNUD/BuildIT/day2/bubbleSort.c
#include <stdio.h>
#include <stdbool.h>

int const length = 16;
int source[length] = {3, 5, 2, 3, 1, 285, 2, 5, 1, 24, 5, 6, 6, 3, 12, 2};

void swap(int indexA, int indexB)
{
    int temp = source[indexB];
    source[indexB] = source[indexA];
    source[indexA] = temp;
}

void bubbleSort()
{
    while (true)
    {
        int counter = 0;

        for (int i = 0; i < length - 1; i++)
        {
            if (source[i] > source[i + 1])
            {
                swap(i, i + 1);
                counter++;
            }
        }

        if (counter == 0)
        {
            return;
        }
    }
}

int main(int argc, char const *argv[])
{
    printf("Int size is %d \n", (int)sizeof(int));
    printf("Array size is %d \n", (int)sizeof(source));

    printf("Source: \n");
    for (int i = 0; i < length; i++)
    {
        printf("%d ", source[i]);
    }

    bubbleSort();

    printf("Sorted: \n");
    for (int i = 0; i < length; i++)
    {
        printf("%d ", source[i]);
    }

    printf("\n");
}
